File: LinkedList/linkedlist.js
var LinkedList = function() {
  this.head = null;
  this.tail = null;
};
var MakeNode = function(value) {
  this.value = value;
  this.next = null;
};

LinkedList.prototype.unshift = function(value) {
  var node = new MakeNode(value);
  if (this.head) {
    node.next = this.head;
    this.head = node;
  } else {
    this.head = node;
    this.tail = node;
  }
};
LinkedList.prototype.shift = function() {
  var temp = this.head;
  this.head = this.head.next;
  return temp;
};
LinkedList.prototype.push = function(value) {
  var node = new MakeNode(value);
  if (this.tail) {
    this.tail.next = node;
    this.tail = node;
  } else {
    this.head = node;
    this.tail = node;
  }
};
LinkedList.prototype.pop = function(node, temp) {
  node = node || this.head;
  temp = this.tail;
  if (node.next === this.tail) {
    this.tail = node;
    this.tail.next = null;
  } else if (node.next) {
    return this.pop(node.next);
  }
  return temp;
};

LinkedList.prototype.contains = function(value, node) {
  node = node || this.head;
  if (node.value === value) {
    return true;
  } else if (node.next) {
    return this.contains(value, node.next);
  }
  return false;
};

File: LinkedList/linkedlist.rb
class LinkedList
  attr_writer :head, :tail

  def initialize
    @head, @tail = nil, nil

  end

  def unshift(value)
    node = Node.new(value)
    if @head
      node.next = @head
      @head = node
    else
      @tail, @head = node, node
    end

  end

  def shift
    temp = @head
    @head = @head.next
    return temp
  end

  def push(value)
    node = Node.new(value)
    if @tail
      @tail.next = node
      @tail = node
    else
      @head, @tail = node, node
    end

  end

  def pop(node=@head)
    if node.next == @tail
      temp = @tail.next
      @tail = node
      @tail.next = nil
      return temp
    elsif node.next
      self.pop(node.next)
    end

  end

  def contains(value, node=@head)
    if node.value == value
      return true
    elsif node.next
      return node.contains(value, node.next)
    end
    return false

  end

  private
  class Node
    attr_accessor :value, :next

    def initialize(value)
      @value = value
      @next = nil
    end

  end

end
